add ltc2672 and ltc2664

kseerp opened the pull request:
LTC2664 4 channel, 16 bit Voltage Output SoftSpan DAC
LTC2672 5 channel, 16 bit Current Output Softspan DAC

since ltc2664 and ltc2672 have the same registers/softspan, thus leveraging the use of this PR https://github.com/analogdevicesinc/linux/pull/2408 and there's plenty of shared codes

i'll plan to utilize also the automatic cleanup for mutex but my current kernel used does not yet support that

nunojsa commented on Documentation/devicetree/bindings/iio/dac/adi,ltc2664.yaml:
```diff
@@ -0,0 +1,210 @@
+title: Analog Devices LTC2664 and LTC2672 DAC
+
+maintainers:
+  - Kim Seer Paller <kimseer.paller@analog.com>
```
you should likely add Michael in here... And have a `Co-developed-by:` tag in the commit message

nunojsa commented on Documentation/devicetree/bindings/iio/dac/adi,ltc2664.yaml:
```diff
@@ -0,0 +1,210 @@
+required:
+  - compatible
+  - reg
+  - spi-max-frequency
```
vcc and iovcc should be required

nunojsa commented on Documentation/devicetree/bindings/iio/dac/adi,ltc2664.yaml:
```diff
@@ -0,0 +1,210 @@
+            adi,output-current-microamp:
+              description: Specify the channel output full scale range.
+              $ref: /schemas/types.yaml#/definitions/uint32
+              enum: [3125000, 6250000, 12500000, 25000000, 50000000, 100000000, 200000000, 300000000]
```
isn't this a big line?

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+	st = iio_priv(indio_dev);
+	st->spi = spi;
+
+	chip_info = &ltc2664_chip_info_tbl[spi_get_device_id(spi)->driver_data];
```
`spi_get_device_match_data()`

kseerp replied:
When using the `spi_get_device_match_data()` helper function, the `const` qualifier is applied. This results in a build warning when the` const` keyword is also assigned to chip_info. The warning may be due to modifications made to the iio_chan fields with ext_info, info_mask_separate, type during device initialization. 

What do you suggest we do, should we retain the initial implementation?

nunojsa replied:
Hmm, the you need to think again how to use chip_info :). That is something that should be `const` (from a logical point of view) and it should serve as warning to you (as a not so good pattern) to rethink what you're doing :smile: . Anyways, I'll elaborate more when reviewing.

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+
+	ret = regulator_bulk_enable(ARRAY_SIZE(st->regulators), st->regulators);
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to enable regulators\n");
```
`devm_regulator_bulk_get_enabled()` - (when upstreaming - not sure if we have it in our tree already)

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+		break;
+	default:
+		return -EINVAL;
+	}
```
I would not go under this trouble... Drop the switch case. The expectation is that people validate their bindings and that should already tell them `adi,manual-span-operation-config` is not supported for ltc2672. Or just have checks for the id after reading the property (but I would likely also skip that).

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+		ret = device_property_read_u32(dev, "adi,manual-span-operation-config", &mspan);
+		if (!ret) {
+			if (mspan > ARRAY_SIZE(ltc2664_mspan_lut)) {
+				fwnode_handle_put(child);
```
why? copy paste?

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+		ret = device_property_read_u32(dev, "adi,rfsadj-ohms", &st->rfsadj);
+		if (!ret) {
+			if (st->rfsadj < 19000 || st->rfsadj > 41000) {
+				fwnode_handle_put(child);
```
ditto...

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+		}
+
+		if (reg >= chip_info->num_channels) {
+			fwnode_handle_put(child);
```
when upstreaming check the new `device_for_each_child_node_scoped()` macro (I think that's the name). It will allow you to skip all fwnode_handle_put() calls.

kseerp replied:
Is that in v6.8, if i'm right.? i'll update my rpi for that so I can also used the cleanup for mutex

nunojsa replied:
The cleanup is in v6.8 - so mutex automatic release is already available . The device_for_each_child_node_scoped() should only land in 6.9-rc1 (next sunday) but you should be able to use it as Jonathan will likely rebase the IIO testing branch soon enough after -rc1

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+				}
+
+				chan->span = span;
+			} else {
```
this could also mean `ret` is an error code. Do we still want to have `chan->span = ltc2664_mspan_lut[mspan][0];` in that case?

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+			}
+			break;
+		default:
+			return -EINVAL;
```
I would again refactor to drop the switch case...

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+		.num_channels = 5,
+		.iio_chan = ltc2672_channels,
+	},
+};
```
please drop the array and have two ltc2664_chip_info variables... Also I guess it can be `const`

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+	LTC2664_CHANNEL(2),
+	LTC2664_CHANNEL(3),
+	LTC2664_CHANNEL(4),
+};
```
not `const`?

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+	LTC2664_CHANNEL(1),
+	LTC2664_CHANNEL(2),
+	LTC2664_CHANNEL(3),
+};
```
ditto

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+			if (max == ltc2672_span_helper[span])
+				return span;
+		}
+		break;
```
Hmm just make the span_helper array part of the chip_info. Then you should be able to simplify this function

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+	if (ret)
+		return ret;
+
+	mutex_lock(&st->lock);
```
use the new cleanup magic. Check  cleanup.h. Basically you do `guard(mutex)(&st->lock)` and "forget" about `mutex_unlock()`. Then, you can return "in place"

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+			*val = ltc2672_span_helper[span] / 250;
+		else
+			*val = 0;
+		break;
```
Likely you can also simplify this with a chip_info span_helper

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+		else
+			*val = LTC2672_SCALE_MULTIPLIER(span) *
+			       (1000 * st->vref / st->rfsadj);
+		break;
```
do `return 0` already in here. Same above

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+	struct ltc2664_chan channels[LTC2672_MAX_CHANNEL];
+	/* lock to protect against multiple access to the device and shared data */
+	struct mutex lock;
+	struct ltc2664_chip_info *chip_info;
```
not `const`?

nunojsa replied:
make this `const`

nunojsa reviewed (commented):
You're also not updating the MAINTAINERS file. Again, make sure to add Michael as someone how also developed significantly the driver (with a Co-developed-by; tag)

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+struct ltc2664_chip_info {
+	char *name;
+	unsigned int num_channels;
+	struct iio_chan_spec *iio_chan;
```
so., make `iio_chan` `const`. This should have the constant, default channel specification. Then move this "dynamic" variable to your `ltc2664_state` struct

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+	.ext_info = ltc2664_ext_info,					\
+}
+
+static struct iio_chan_spec ltc2664_channels[] = {
```
should be `const` now

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+	LTC2664_CHANNEL(3),
+};
+
+static struct iio_chan_spec ltc2672_channels[] = {
```
ditto

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+	LTC2664_CHANNEL(4),
+};
+
+static struct ltc2664_chip_info ltc2664_chip_info_tbl[] = {
```
and ditto

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,831 @@
+	chip_info->iio_chan = devm_kmemdup(&st->spi->dev, chip_info->iio_chan,
+					   chip_info->num_channels *
+					   sizeof(*chip_info->iio_chan),
+					   GFP_KERNEL);
```
Now in here you duplicate your `const` data and assign it to `st->channels` so it may be changed. See the difference? Whit this, you should be fine in using `spi_get_device_match_data()` down on probe...

nunojsa reviewed (commented):
Some more comments as it makes it easier to explain my points about `chip_info`

kseerp commented:
appreciate all the reviews, here are the changelogs for v2 and changes have been tested on relevant hardware

- vcc and iovcc added as required property
- updated MAINTAINERS file
- utilized cleanup for mutex automatic release
- added span_helper and num_span in chip_info, and restructured the ltc2672_span_helper into a 2d array to simplify the span_lookup function (uncertain if I implemented it properly or in standard implementation)
- refactored offset and scale function
- made iio_chan constant for default channel specification and moved dynamic variable to ltc2664_state struct
- dropped switch case and refactored channel_config function
- utilized spi_get_device_match_data()
- added Co-developed-by tag

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,798 @@
+
+	st->rfsadj = 20000;
+	ret = device_property_read_u32(dev, "adi,rfsadj-ohms", &st->rfsadj);
+	if (!ret) {
```
After a second thought, I think (to be on the safe side) having a second check on the device id would make sense... So, we make sure no one skips bindings validations and uses properties o devices that don't support them . I would just another `if()` before checking  the property values. Ditto for all specific properties...

nunojsa commented on drivers/iio/dac/ltc2664.c:
```diff
@@ -0,0 +1,798 @@
+		if (ret)
+			break;
+		st->global_toggle = en;
+		break;
```
I would just `return len` in here..

kseerp commented:
added accepted upstream version, modified for compliance with the ADI kernel
